File: apps/mobile/test/features/events/domain/services/private_venue_policy_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:tribely/src/features/events/domain/services/private_venue_policy.dart';

void main() {
  // ---------------------------------------------------------------------------
  // Public categories — all 12 should return not-private with a clean name.
  // ---------------------------------------------------------------------------
  group('public categories + clean name → not private', () {
    const publicCategories = [
      'hawker_centre',
      'park',
      'museum',
      'restaurant',
      'bar',
      'cafe',
      'beach',
      'mrt_landmark',
      'library',
      'community_centre',
      'shopping_mall_common_area',
      'tourist_attraction',
    ];

    for (final category in publicCategories) {
      test('$category → not private', () {
        final result = detectPrivateVenue(
          categoryValue: category,
          venueName: 'Lau Pa Sat',
        );
        expect(result.isPrivate, isFalse);
        expect(result.reason, isNull);
        expect(result.matchedKeyword, isNull);
      });
    }
  });

  // ---------------------------------------------------------------------------
  // Private categories — all 6 should return categoryNotPublic.
  // ---------------------------------------------------------------------------
  group('private categories + clean name → categoryNotPublic', () {
    const privateCategories = [
      'apartment',
      'condo',
      'home',
      'hotel',
      'hostel',
      'other',
    ];

    for (final category in privateCategories) {
      test('$category → categoryNotPublic', () {
        final result = detectPrivateVenue(
          categoryValue: category,
          venueName: 'Lau Pa Sat',
        );
        expect(result.isPrivate, isTrue);
        expect(result.reason, PrivateVenueReason.categoryNotPublic);
        expect(result.matchedKeyword, isNull);
      });
    }
  });

  // ---------------------------------------------------------------------------
  // Null category
  // ---------------------------------------------------------------------------
  group('null category', () {
    test('null category + clean name → not private', () {
      final result = detectPrivateVenue(
        categoryValue: null,
        venueName: 'Gardens by the Bay',
      );
      expect(result.isPrivate, isFalse);
      expect(result.reason, isNull);
    });

    test('null category + "my apartment" → keywordMatch (apartment)', () {
      final result = detectPrivateVenue(
        categoryValue: null,
        venueName: 'my apartment',
      );
      expect(result.isPrivate, isTrue);
      expect(result.reason, PrivateVenueReason.keywordMatch);
      expect(result.matchedKeyword, 'apartment');
    });
  });

  // ---------------------------------------------------------------------------
  // Every keyword — positive case proving case-insensitive contains match.
  // ---------------------------------------------------------------------------
  group('keyword matches (null category, case-insensitive contains)', () {
    const keywordCases = [
      ('condominium', 'Sunrise Condominium Tower'),
      ('apartment', 'City apartment on Level 3'),
      ('my place', 'Come to my place tonight'),
      ('my flat', 'Gathering at my flat'),
      ('my room', 'Chill at my room'),
      ('airbnb', 'Airbnb unit near Orchard'),
      ('studio', 'Studio near Bugis'),
      ('hostel', 'Capsule Hostel Dormitory'),
      ('condo', 'The Pinnacle Condo'),
      ('hotel', 'Grand Hotel lobby'),
      ('motel', 'Budget Motel near airport'),
      ('house', 'Heritage house on Emerald Hill'),
      ('home', 'home base for the group'),
      ('unit', 'unit #04-12 Tanjong Pagar'),
      ('apt', 'Apt 5B Marine Parade'),
    ];

    for (final (keyword, name) in keywordCases) {
      test('"$keyword" detected in "$name"', () {
        final result = detectPrivateVenue(
          categoryValue: null,
          venueName: name,
        );
        expect(result.isPrivate, isTrue);
        expect(result.reason, PrivateVenueReason.keywordMatch);
        expect(result.matchedKeyword, keyword);
      });
    }
  });

  // ---------------------------------------------------------------------------
  // HDB false-positive guard — "Block" must NOT trigger a match.
  // ---------------------------------------------------------------------------
  group('HDB address false-positive guard', () {
    test('"Block 123 Smith St" with public category → not private', () {
      final result = detectPrivateVenue(
        categoryValue: 'hawker_centre',
        venueName: 'Block 123 Smith Street',
      );
      expect(result.isPrivate, isFalse);
    });

    test('"Block 335 Bukit Timah" with null category → not private', () {
      final result = detectPrivateVenue(
        categoryValue: null,
        venueName: 'Block 335 Bukit Timah',
      );
      expect(result.isPrivate, isFalse);
    });
  });

  // ---------------------------------------------------------------------------
  // Case-insensitivity
  // ---------------------------------------------------------------------------
  group('case-insensitive matching', () {
    test('"MY APARTMENT" matches apartment keyword', () {
      final result = detectPrivateVenue(
        categoryValue: null,
        venueName: 'MY APARTMENT',
      );
      expect(result.isPrivate, isTrue);
      expect(result.reason, PrivateVenueReason.keywordMatch);
      expect(result.matchedKeyword, 'apartment');
    });

    test('"HOSTEL" matches hostel keyword', () {
      final result = detectPrivateVenue(
        categoryValue: null,
        venueName: 'HOSTEL',
      );
      expect(result.isPrivate, isTrue);
      expect(result.reason, PrivateVenueReason.keywordMatch);
      expect(result.matchedKeyword, 'hostel');
    });
  });

  // ---------------------------------------------------------------------------
  // Priority — category check wins over keyword match.
  // ---------------------------------------------------------------------------
  group('priority: category check wins over keyword match', () {
    test('category=apartment + name="My Apartment" → categoryNotPublic', () {
      final result = detectPrivateVenue(
        categoryValue: 'apartment',
        venueName: 'My Apartment',
      );
      expect(result.isPrivate, isTrue);
      expect(result.reason, PrivateVenueReason.categoryNotPublic);
      // matchedKeyword must be null — category check short-circuits keyword scan.
      expect(result.matchedKeyword, isNull);
    });

    test('category=home + name="Condo near MRT" → categoryNotPublic', () {
      final result = detectPrivateVenue(
        categoryValue: 'home',
        venueName: 'Condo near MRT',
      );
      expect(result.isPrivate, isTrue);
      expect(result.reason, PrivateVenueReason.categoryNotPublic);
      expect(result.matchedKeyword, isNull);
    });
  });

  // ---------------------------------------------------------------------------
  // Greedy / longest-first matching — "condominium" wins over "condo".
  // ---------------------------------------------------------------------------
  group('longest-first greedy matching', () {
    test('"Sunrise Condominium" matches condominium not condo', () {
      final result = detectPrivateVenue(
        categoryValue: null,
        venueName: 'Sunrise Condominium',
      );
      expect(result.isPrivate, isTrue);
      expect(result.matchedKeyword, 'condominium');
    });
  });

  // ---------------------------------------------------------------------------
  // Equatable contract.
  // ---------------------------------------------------------------------------
  group('PrivateVenueDetection equality', () {
    test('identical detections are equal', () {
      const a = PrivateVenueDetection(
        isPrivate: true,
        reason: PrivateVenueReason.keywordMatch,
        matchedKeyword: 'condo',
      );
      const b = PrivateVenueDetection(
        isPrivate: true,
        reason: PrivateVenueReason.keywordMatch,
        matchedKeyword: 'condo',
      );
      expect(a, equals(b));
    });

    test('different matchedKeyword → not equal', () {
      const a = PrivateVenueDetection(
        isPrivate: true,
        reason: PrivateVenueReason.keywordMatch,
        matchedKeyword: 'condo',
      );
      const b = PrivateVenueDetection(
        isPrivate: true,
        reason: PrivateVenueReason.keywordMatch,
        matchedKeyword: 'hotel',
      );
      expect(a, isNot(equals(b)));
    });
  });
}

File: apps/api/src/features/events/domain/services/private-venue-policy.test.ts
import { describe, expect, it } from 'vitest';
import { VenueCategory } from '../value-objects/venue-category.js';
import { PRIVATE_KEYWORDS, detectPrivateVenue } from './private-venue-policy.js';

describe('detectPrivateVenue', () => {
  describe('public category + clean name → not private', () => {
    it.each(VenueCategory.PUBLIC_VALUES)(
      'returns isPrivate=false for public category "%s" with a clean name',
      (value) => {
        const result = detectPrivateVenue({
          category: VenueCategory.create(value),
          venueName: 'Gardens by the Bay',
        });
        expect(result).toEqual({ isPrivate: false, reason: null, matchedKeyword: null });
      },
    );
  });

  describe('private category → category_not_public', () => {
    const privateValues = VenueCategory.VALUES.filter(
      (v) => !(VenueCategory.PUBLIC_VALUES as readonly string[]).includes(v),
    );

    it.each(privateValues)(
      'returns category_not_public for private category "%s"',
      (value) => {
        const result = detectPrivateVenue({
          category: VenueCategory.create(value),
          venueName: 'Some Venue',
        });
        expect(result).toEqual({
          isPrivate: true,
          reason: 'category_not_public',
          matchedKeyword: null,
        });
      },
    );
  });

  describe('keyword matching — case-insensitive substring', () => {
    it.each(PRIVATE_KEYWORDS)(
      'matches keyword "%s" in lowercase',
      (keyword) => {
        const result = detectPrivateVenue({
          category: VenueCategory.create('park'),
          venueName: `near the ${keyword} entrance`,
        });
        expect(result).toEqual({
          isPrivate: true,
          reason: 'keyword_match',
          matchedKeyword: keyword,
        });
      },
    );

    it.each(PRIVATE_KEYWORDS)(
      'matches keyword "%s" in UPPERCASE',
      (keyword) => {
        const result = detectPrivateVenue({
          category: VenueCategory.create('park'),
          venueName: keyword.toUpperCase(),
        });
        expect(result).toEqual({
          isPrivate: true,
          reason: 'keyword_match',
          matchedKeyword: keyword,
        });
      },
    );

    it.each(PRIVATE_KEYWORDS)(
      'matches keyword "%s" in mixed case',
      (keyword) => {
        // Toggle case of each character for a mixed-case variant
        const mixed = keyword
          .split('')
          .map((ch, i) => (i % 2 === 0 ? ch.toUpperCase() : ch.toLowerCase()))
          .join('');
        const result = detectPrivateVenue({
          category: VenueCategory.create('park'),
          venueName: `My ${mixed} nearby`,
        });
        expect(result).toEqual({
          isPrivate: true,
          reason: 'keyword_match',
          matchedKeyword: keyword,
        });
      },
    );
  });

  describe('HDB false-positive guard', () => {
    it('does NOT flag "block" in a typical HDB address', () => {
      const result = detectPrivateVenue({
        category: VenueCategory.create('hawker_centre'),
        venueName: 'Block 335 Smith Street Hawker Centre',
      });
      expect(result).toEqual({ isPrivate: false, reason: null, matchedKeyword: null });
    });
  });

  describe('priority: category_not_public wins over keyword match', () => {
    it('returns category_not_public when category is private AND name contains a keyword', () => {
      // category=apartment (private) + name contains "apartment" keyword
      const result = detectPrivateVenue({
        category: VenueCategory.create('apartment'),
        venueName: 'My Apartment block 3',
      });
      expect(result).toEqual({
        isPrivate: true,
        reason: 'category_not_public',
        matchedKeyword: null,
      });
    });
  });

  describe('specific case-insensitive spot-checks', () => {
    it('"MY APARTMENT" matches apartment keyword', () => {
      const result = detectPrivateVenue({
        category: VenueCategory.create('park'),
        venueName: 'MY APARTMENT',
      });
      expect(result).toEqual({
        isPrivate: true,
        reason: 'keyword_match',
        matchedKeyword: 'apartment',
      });
    });

    it('"My ApArTmEnT" matches apartment keyword', () => {
      const result = detectPrivateVenue({
        category: VenueCategory.create('park'),
        venueName: 'My ApArTmEnT',
      });
      expect(result).toEqual({
        isPrivate: true,
        reason: 'keyword_match',
        matchedKeyword: 'apartment',
      });
    });
  });
});
